tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -4037,41 +4042,38 @@ Status ConvertBatchMatMul(OpConverterParams* params) {
   const bool transpose_a = attrs.get<bool>("adj_x");
   const bool transpose_b = attrs.get<bool>("adj_y");
 
-  // Removes the batch dimension from weights.
-  const auto remove_weights_batch_dim =
-      [&params](const TRT_TensorOrWeights& input, TRT_TensorOrWeights* tensor) {
-        auto dims = input.GetTrtDims();
-        if (input.is_weights()) {
-          // The other operand must be a tensor, this is ensured by earlier
-          // checks. Checks that the batch dimension is not changed by
-          // broadcasting.
-          if (dims.d[0] != 1) {
-            return errors::InvalidArgument(
-                "Input weight attempts to broadcast across batch dimension for "
-                "BatchMatMul, at ",
-                params->node_def.name());
-          }
-          // Remove the batch dimension from the weights.
-          TF_RETURN_IF_ERROR(RemoveBatchDimension(&dims));
-        }
-        // Create tensor and reshape if necessary.
-        nvinfer1::ITensor* t{nullptr};
-        TF_RETURN_IF_ERROR(params->converter->PrepareTensorForShape(
-            input, dims, params->validation_only, &t));
-        *tensor = TRT_TensorOrWeights{t};
-        return Status::OK();
-      };
-
-  TRT_TensorOrWeights tensor_l{nullptr};
-  TRT_TensorOrWeights tensor_r{nullptr};
-  TF_RETURN_IF_ERROR(remove_weights_batch_dim(inputs.at(0), &tensor_l));
-  TF_RETURN_IF_ERROR(remove_weights_batch_dim(inputs.at(1), &tensor_r));
-
-  if (params->validation_only) {
+  // There is no way to batch constants in TRT. Example:
+  // Tensor with TF Dims: 12 5 3 -> TRT Dims: 5 3
+  // Weight with TF Dims: 12 3 6 -> TRT Dims: 12 3 6
+  // It is not possible to treat the weight input as a batched [3, 6] tensor.
+  const auto check_weight_is_not_batched = [](
+      const TRT_TensorOrWeights& input_l, const TRT_TensorOrWeights& input_r) {
+    if (input_l.is_weights() &&
+        input_l.GetTrtDims().nbDims > input_r.GetTrtDims().nbDims &&
+        input_l.GetTrtDims().d[0] != 1) {
+      return errors::Unimplemented(
+          "TensorRT does not support batched constants.");
+    }
     return Status::OK();
-  }
+  };
+  TF_RETURN_IF_ERROR(check_weight_is_not_batched(inputs.at(0), inputs.at(1)));
+  TF_RETURN_IF_ERROR(check_weight_is_not_batched(inputs.at(1), inputs.at(0)));
 
-  return ConvertMatMulHelper(params, tensor_l, tensor_r, transpose_a,
+  // Broadcast inputs.
+  nvinfer1::Dims broadcasted_dims_l, broadcasted_dims_r;
+  TF_RETURN_IF_ERROR(GetTrtBroadcastShape(
+      inputs.at(0), inputs.at(1), /*check_feasibility=*/false,
```

Comment:
For typically uses of broadcast, we checked that the input dimensions either matched exactly or one of them was 1.
But for a batch matrix multiply, the dimensions don't line up exactly:
```
input 0: [N, T, C]
input 1: [1, C, K]
output: [N, T, K]
```
Since `C != K` and `T != C`, checkfeasiblity would fail